### PR TITLE
fix micmute led, init smart charge

### PR DESCRIPTION
**Pull request by sermart1234**
experimental support smart charge

### CODE REVIEW

**Review by aymanbagabas — COMMENTED** [review]
Good job @sermart1234 this is looking good!

File: huawei-wmi.c [hunk 1]
```diff
@@ -391,7 +397,8 @@ static void huawei_wmi_leds_setup(struct device *dev)
 	huawei->cdev.dev = dev;
 	huawei->cdev.flags = LED_CORE_SUSPENDRESUME;
 
-	devm_led_classdev_register(dev, &huawei->cdev);
+	if (acpi_has_method(NULL, "\\SMLS") || (quirks && quirks->ec_micmute))
+		devm_led_classdev_register(dev, &huawei->cdev);
```
Comment (aymanbagabas) [inline huawei-wmi.c:401]: Nice work checking `\SMLS`

This function `huawei_wmi_leds_setup` should handle all of the device LEDs micmute, Fn, etc. Perhaps change the `cdev` led name to `micmute_cdev`. Also I would cover the whole micmute led block in a if statement and implement a `huawei_wmi_leds_exit` function to unregister the leds.

```
huawei->micmute_led_available = false;
if (acpi_has_method("\\SMLS") || ...) {
  huawei->micmute_led_available = true;
  huawei->micmute_cdev.name = "platform::micmute";
  // ...
  // register led
}
// other less
```

File: huawei-wmi.c [hunk 2]
```diff
@@ -1173,7 +1359,16 @@ static int read_ec_by_name(char *acpi_method, u64 *value)
 	return -ENODEV;
 }
 
+
+
 /*FN led (read only)*/
+
+static void huawei_wmi_led_fake_set(struct led_classdev *led_cdev,
```
[the remark this answers is not in the document]
Comment (sermart1234) [reply huawei-wmi.c:1366]: Led read-only. Without this, I get an error message in dmesg
```
leds platform::fn_led: Setting an LED's brightness failed (-524)
leds platform::KCMS: Setting an LED's brightness failed (-524)
```

**Comment by mpi3d** [comment]
I can confirm that it works on HUAWEI MateBook D 15 AMD.
Mic led is still working too.
Thanks for your work @sermart1234. @aymanbagabas Thanks to you too.
Any idea when this could be merged in the kernel?

File: huawei-wmi.c (hunk 1 above)
Comment (aymanbagabas) [reply huawei-wmi.c:401]: i would say for that, users should use the `ec_micmute` module param

**Comment by sermart1234** [comment]
@mpi3d @aymanbagabas
1) smart charging test (this is not the final implementation. there may be conflicts with thresholds)
2) finish the LEDs and hwmon
3) I don't like quirks
4) wq branch
5) there are still unrealized WMI methods

**Comment by aymanbagabas** [comment]
> I can confirm that it works on HUAWEI MateBook D 15 AMD. Mic led is still working too. Thanks for your work @sermart1234. @aymanbagabas Thanks to you too. Any idea when this could be merged in the kernel?

We could start merging that into kernel once we thoroughly test these changes and confirm their backward compatibility and coverage across different Huawei models

**Comment by sermart1234** [comment]
@mpi3d Do you have any ideas to improve the smart charging interface?

**Comment by mpi3d** [comment]
Ok, thanks a lot.

**Review by egormanga — APPROVED** [review]
Minor spelling notices. LGTM otherwise.

File: huawei-wmi.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (aymanbagabas) [reply huawei-wmi.c:1366]: can't you have `brightness_set` null? i.e don't set it

**Comment by aymanbagabas** [comment]
> @mpi3d @aymanbagabas
> 
> 1. smart charging test (this is not the final implementation. there may be conflicts with thresholds)
> 2. finish the LEDs and hwmon
> 3. I don't like quirks
> 4. wq branch
> 5. there are still unrealized WMI methods

using Ghidra to reverse engineer Huawei PC Manger will definitely help. You can decompile the DLLs that come with PC Manager and try to understand how it works.

**Comment by aymanbagabas** [comment]
> @aymanbagabas Can I hook atkbd events? Or do I need to patch atkbd to remove these messages in dmesg? It would be nice to get events from hotkeys (keyboard backlight key, display backlight button, power unlock (FN + P)) изображение изображение

The atkbd driver emits these keys. And since huawei-wmi handles key inputs using `linux/input/sparse-keymap`, we can ignore atkbd events and rely on huawei-wmi registered input device to do the right action.

You can ignore atkbd events using [SystemdD keyboard hwdb](https://github.com/systemd/systemd/blob/main/hwdb.d/60-keyboard.hwdb#L759)

**Comment by mpi3d** [comment]
@sermart1234
> @mpi3d Do you have any ideas to improve the smart charging interface?

Sorry, I did not see your message. I don't know, really, what smart charging does. When I activate it, just charge to 100%. Even on Windows. Is it supposed to do something else?

@aymanbagabas
> using Ghidra to reverse engineer Huawei PC Manger will definitely help. You can decompile the DLLs that come with PC Manager and try to understand how it works.

I wish I could help. But I'm already busy with making the fingerprint sensor work. 
Also, I don't know how this work for now. Of course, I can learn, but since you are already working on it, I prefer working on the fingerprint sensor.

**Pull request merged**